File: challenges/web/web-proto-03/app/siem.js
/** Emision de eventos SIEM (esquema ARCHITECTURE.md §5). Fire-and-forget. Port de siem.py. */
'use strict';

const http = require('http');
const { URL } = require('url');

const COLLECTOR_URL = process.env.COLLECTOR_URL || 'http://collector:9000';
const TEAM_ID = process.env.TEAM_ID || 'team_local';
const CHALLENGE_ID = process.env.CHALLENGE_ID || 'web-proto-03';

function _post(event) {
  try {
    const data = Buffer.from(JSON.stringify(event), 'utf-8');
    const u = new URL(COLLECTOR_URL + '/event');
    const req = http.request(
      {
        hostname: u.hostname,
        port: u.port || 80,
        path: u.pathname,
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Content-Length': data.length },
        timeout: 2000,
      },
      (res) => { res.on('data', () => {}); res.on('end', () => {}); }
    );
    req.on('error', () => {});
    req.on('timeout', () => req.destroy());
    req.write(data);
    req.end();
  } catch (e) {
    /* fire-and-forget: nunca tumbar el reto */
  }
}

function emit(eventType, severity, srcIp = null, detail = null) {
  const event = {
    ts: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
    source: 'challenge',
    team_id: TEAM_ID,
    user: TEAM_ID,
    src_ip: srcIp,
    event_type: eventType,
    severity: severity,
    challenge_id: CHALLENGE_ID,
    detail: detail || {},
  };
  _post(event);
}

module.exports = { emit };

File: challenges/web/web-creditview/app/siem.py
"""Emision de eventos al SIEM collector desde un reto.

Respeta el esquema EXACTO de ARCHITECTURE.md seccion 5. Fire-and-forget:
si el collector esta caido, NUNCA rompe la request del jugador.

Copiable tal cual a cualquier reto Python (web/api/crypto).
"""
import logging
import os
import threading
from datetime import datetime, timezone

import urllib.request
import json

logger = logging.getLogger("siem")

COLLECTOR_URL = os.environ.get("COLLECTOR_URL", "http://collector:9000")
TEAM_ID = os.environ.get("TEAM_ID", "team_local")
CHALLENGE_ID = os.environ.get("CHALLENGE_ID", "unknown")


def _post(event: dict) -> None:
    try:
        data = json.dumps(event).encode()
        req = urllib.request.Request(
            f"{COLLECTOR_URL}/event",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        urllib.request.urlopen(req, timeout=2.0)
    except Exception as exc:  # noqa: BLE001 - el collector caido no afecta el juego
        logger.warning("No se pudo enviar evento SIEM: %s", exc)


def emit(event_type: str, severity: str, src_ip: str | None = None, detail: dict | None = None) -> None:
    """Emite un evento async (thread) para no bloquear la respuesta del reto."""
    event = {
        "ts": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source": "challenge",
        "team_id": TEAM_ID,
        "user": TEAM_ID,
        "src_ip": src_ip,
        "event_type": event_type,
        "severity": severity,
        "challenge_id": CHALLENGE_ID,
        "detail": detail or {},
    }
    threading.Thread(target=_post, args=(event,), daemon=True).start()
